org.lflang/src/org/lflang/generator/ReactionInstance.java

```diff
@@ -351,6 +351,7 @@ public Set<ReactionInstance> dependsOnReactions() {
     public Set<Integer> getLevels() {
         Set<Integer> result = new LinkedHashSet<>();
         // Force calculation of levels if it has not been done.
+        // FIXME: Is it necessary to repeat this everywhere?
```

Comment:
@edwardalee Are there any valid reason for all the time checking if the levels have been assigned? This check is performed for each reaction in the program. I am not performing similar test on the deadline assignments, I assume that it is done for the top-level once before code-generating any priorities/indices for the reactions

Reply:
I'm not sure I understand the question, but it looks like getLevels() is called only once per ReactorInstance. The assignLevels() function caches the result, so it doesn't recompute it.  I guess we could cache the result of getLevels(), but I don't think it will change anything currently.

Reply:
I am referring to calling `assignLevels()` all the time. Either it is redundant (because it is invoked on the top-level reactor at the beginning of code generation) or somehow there are scenarios where we are inside the code generator and for some reason levels have not been assigned yet. It is important because in the latter scenario, also the deadlines will not have been assigned. So I would have to replicate that caching logic for the deadlines also.